test/e2e/vmagent_test.go

```diff
@@ -0,0 +1,327 @@
+		if err := wait.PollImmediate(10*time.Second, 1*time.Minute, func() (done bool, err error) {
+			var tempVm vmInfo
+			if vm.osType == "Linux" {
+				tempVm = getVMInfo(t, data, vm.nodeName)
+			} else {
+				tempVm = getWindowsVMInfo(t, data, vm.nodeName)
+			}
+			if !assert.Equal(t, vm.ifName, tempVm.ifName, "Unexpected uplink interface name") {
+				return false, nil
+			}
+			if !assert.Equal(t, vm.ip, tempVm.ip, "Unexpected uplink interface IP") {
+				return false, nil
+			}
+			return true, nil
+		}); err != nil {
+			assert.Fail(t, "Failed to verify uplink configuration after cleanup")
+			return
+		}
+	}
+	t.Logf("TestVMAgent teardown")
+	for _, vm := range vmList {
+		eeName := generateExternalEntityName(vm)
+		if err := data.crdClient.CrdV1alpha1().ExternalNodes(nameSpace).Delete(context.TODO(), vm.nodeName, metav1.DeleteOptions{}); err != nil {
+			t.Logf("Failed to delete externalnode, err: %v", err)
+		}
+		verifyExternalEntityExistence(t, data, eeName, vm.nodeName, false)
+		verifyUpLinkAfterCleanup(vm)
+	}
+}
+
+func verifyExternalEntityExistence(t *testing.T, data *TestData, eeName string, vmNodeName string, expectExists bool) {
+	if err := wait.PollImmediate(10*time.Second, 1*time.Minute, func() (done bool, err error) {
+		t.Logf("Verifying ExternalEntity %s, expectExists %t", eeName, expectExists)
+		_, err = data.crdClient.CrdV1alpha2().ExternalEntities(nameSpace).Get(context.TODO(), eeName, metav1.GetOptions{})
+		if err != nil && !errors.IsNotFound(err) {
+			t.Errorf("Failed to get ExternalEntity %s by ExternalNode %s: %v", eeName, vmNodeName, err)
+			return false, err
+		}
+
+		if expectExists {
+			if err != nil {
+				return false, err
+			}
+			return true, nil
+		}
+		// ExternalEntity is expected to be deleted, check for error
+		if err != nil {
+			return true, nil
+		}
+		return false, nil
+
+	}); err != nil {
+		op := "created"
+		if !expectExists {
+			op = "deleted"
+		}
+		assert.NoError(t, err, "Failed to verify ExternalEntity %s %s by ExternalNode %s", eeName, op, vmNodeName)
+	}
+}
+
+func testExternalNode(t *testing.T, data *TestData, vmList []vmInfo) {
+	verifyExternalNodeRealization := func(vm vmInfo) {
+		if err := wait.PollImmediate(10*time.Second, 1*time.Minute, func() (done bool, err error) {
+			t.Logf("Verify host interface configuration for VM: %s", vm.nodeName)
+			exists, err := verifyInterfaceIsInOVS(t, data, vm)
+			return exists, err
+		}); err != nil {
+			require.NoError(t, err, "Failed to verify host interface in OVS, vmInfo %+v", vm)
+			return
+		}
+
+		var tempVm vmInfo
+		if vm.osType == "Windows" {
+			tempVm = getWindowsVMInfo(t, data, vm.nodeName)
+		} else {
+			tempVm = getVMInfo(t, data, vm.nodeName)
+		}
+		assert.Equal(t, vm.ifName, tempVm.ifName, "Failed to verify uplink interface, Expected %s, Got %s", vm.ifName, tempVm.ifName)
+		assert.Equal(t, vm.ip, tempVm.ip, "Failed to verify uplink IP, Expected %s, Got %s", vm.ip, tempVm.ip)
```

Comment:
`assert.Equal` will dump expected value and actual value when the check fails, so you can remove "Expected %s, Got %s" from the message.

Reply:
Done